File: src/openroad_mcp/utils/path_security.py
"""Path security validation utilities to prevent directory traversal attacks."""

from pathlib import Path

from ..core.exceptions import ValidationError


def validate_path_segment(segment: str, segment_name: str) -> None:
    """Validate a path segment to prevent directory traversal attacks.

    Rejects empty strings, '.', '..', path separators, null bytes,
    and glob characters (*, ?, [, ]).
    """
    if not segment:
        raise ValidationError(f"{segment_name} cannot be empty")

    if segment in (".", ".."):
        raise ValidationError(f"{segment_name} cannot be '.' or '..'")

    if "/" in segment or "\\" in segment:
        raise ValidationError(f"{segment_name} cannot contain path separators")

    if "\x00" in segment:
        raise ValidationError(f"{segment_name} cannot contain null bytes")

    if any(char in segment for char in "*?[]"):
        raise ValidationError(f"{segment_name} cannot contain glob characters (* ? [ ])")


def validate_safe_path_containment(target_path: Path, base_path: Path, context: str) -> None:
    """Validate that resolved target path is safely contained within base path.

    Uses Path.resolve() to handle symlinks and relative paths, then verifies
    the resolved target is a child of the base path using relative_to().
    """
    try:
        resolved_target = target_path.resolve()
        resolved_base = base_path.resolve()
    except (OSError, RuntimeError) as e:
        raise ValidationError(f"Failed to resolve {context} path: {e}") from e

    try:
        resolved_target.relative_to(resolved_base)
    except ValueError as e:
        raise ValidationError(f"{context} path {target_path} is not contained within {base_path}") from e

File: typescript/src/utils/path_security.ts
import path from "node:path";
import fs from "node:fs";
import { ValidationError } from "../exceptions.js";

export function validatePathSegment(segment: string, segmentName: string): void {
  if (!segment) throw new ValidationError(`${segmentName} cannot be empty`);
  if (segment === "." || segment === "..") throw new ValidationError(`${segmentName} cannot be '.' or '..'`);
  if (segment.includes("/") || segment.includes("\\")) throw new ValidationError(`${segmentName} cannot contain path separators`);
  if (segment.includes("\x00")) throw new ValidationError(`${segmentName} cannot contain null bytes`);
  if (/[*?[\]]/.test(segment)) throw new ValidationError(`${segmentName} cannot contain glob characters (* ? [ ])`);
}

export function validateSafePathContainment(targetPath: string, basePath: string, context: string): void {
  let resolvedTarget: string;
  let resolvedBase: string;

  try {
    resolvedBase = fs.realpathSync(basePath);
  } catch (e) {
    throw new ValidationError(`Failed to resolve ${context} path: ${e}`);
  }

  try {
    resolvedTarget = fs.realpathSync(targetPath);
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code !== "ENOENT") {
      throw new ValidationError(`Failed to resolve ${context} path: ${e}`);
    }
    // Walk up to find the longest existing prefix, resolve its symlinks, then
    // re-append the non-existent suffix. A plain path.resolve() is unsafe here
    // because it won't resolve symlinks in existing parent directories, allowing
    // e.g. base/evil_link/nonexistent to escape containment at runtime.
    const suffix: string[] = [];
    let current = path.resolve(targetPath);
    for (;;) {
      const parent = path.dirname(current);
      if (parent === current) {
        resolvedTarget = path.resolve(targetPath);
        break;
      }
      suffix.unshift(path.basename(current));
      current = parent;
      try {
        resolvedTarget = path.join(fs.realpathSync(current), ...suffix);
        break;
      } catch (innerErr) {
        if ((innerErr as NodeJS.ErrnoException).code !== "ENOENT") {
          throw new ValidationError(`Failed to resolve ${context} path: ${innerErr}`);
        }
      }
    }
  }

  const rel = path.relative(resolvedBase, resolvedTarget);
  const firstComponent = rel.split(path.sep)[0];
  if (firstComponent === ".." || path.isAbsolute(rel)) {
    throw new ValidationError(`${context} path ${targetPath} is not contained within ${basePath}`);
  }
}
